lib/internal/process/next_tick.js

```diff
@@ -19,6 +19,69 @@ function setupNextTick() {
   const { kInit, kDestroy, kAsyncIdCounter } = async_wrap.constants;
   const { async_id_symbol, trigger_async_id_symbol } = async_wrap;
 
+  class SmallQueue {
+    constructor() {
+      this.list = [];
+      this.top = 0;
+      this.btm = 0;
+    }
+
+    push(tick) {
+      if (this.top === this.list.length) {
+        this.list.push(tick);
+        this.top++;
+      } else {
+        this.list[this.top++] = tick;
+      }
+    }
+
+    shift() {
+      if (this.btm < this.top) {
+        var next = this.list[this.btm];
+        this.list[this.btm++] = undefined;
+        if (this.top === this.btm) {
+          this.btm = this.top = 0;
+        }
+        return next;
+      }
+      return null;
+    }
+  }
+
+  var tail = null;
+  var head = new SmallQueue();
+
+  function push(tick) {
+    if (tail === null && head.top >= 1024 && head.btm > 0) {
+      tail = head;
+      head = new SmallQueue();
+    }
+
+    if (head.top === 0 && tail === null) {
+      tickInfo[kHasScheduled] = 1;
+    }
+
+    head.push(tick);
+  }
+
+  function shift() {
+    var next;
+
+    if (tail !== null) {
+      next = tail.shift();
+      if (next !== null) return next;
```

Comment:
If I am not mistaken this `if (next !== null)` check could be removed by moving the `if (this.btm < this.top) {` check from `SmallQueue#shift()` in here and to always return a value when calling `SmallQueue#shift()`.

Reply:
```js
if (tail !== null) {
  if (tail.btm < tail.top) return tail.shift();
  tail = null;
}
```